proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -500,20 +511,89 @@ Exposing the property-wrapper storage type through the type system has the follo
+
+```swift
+func hasSideEffect() -> Int {
+  struct S {
+    static var state = 0
+  }
+
+  S.state += 1
+  return S.state
+}
+
+@propertyWrapper
+struct Wrapper {
+  var wrappedValue: Int
+
+  init(wrappedValue: Int) {
+    self.wrappedValue = wrappedValue + hasSideEffect()
+  }
+}
+
+func demonstrateEvaluationOrder(@Wrapper arg1: Int, arg2: Int) {
+  print(arg1, arg2)
+}
+
+demonstrateEvaluationOrder(arg1: 1, arg2: hasSideEffect())
+```
+
+If the property wrapper initializer is evaluated in the caller, the output of this code is `2, 2`. If the property wrapper initializer is evaluated in the callee, the output of the code is `3, 1`.
+
+The proposal authors believe that these two kinds of property wrappers already exist today, and formalizing the distinction is a first step in enhancing programmers' understanding of such a complex feature. Property wrappers are very flexible to cover a wide variety of use cases. Formalizing the two broad categories of use cases opens up many interesting possibilities for the language and compiler to enhance library documentation when API wrappers are used, provide better guidance to programmers, and even allow library authors to augment the guidance given to programmers on invalid code.
+
+### Explicit spelling for API-level property wrappers
+
+The scope of what is considered an API-level property wrapper is very limited in this proposal, and the external effect of an API-level property wrapper may be useful for wrappers that don't fit the current definition. The `@propertyWrapper` attribute could have an explicit `apiLevel` option that allows library authors to define whether the property wrapper has an external effect on the wrapped declaration:
+
+```swift
+@propertyWrapper(apiLevel)
+struct Asserted<Value> {
+  init(
+    wrappedValue: Value,
+    _ assertion: (Value) -> Bool,
+    file: StaticString = #file,
+    line: UInt = #line
+  ) { ... }
+}
+```
+
+### Library-defined diagnostic notes for understanding API wrapper restrictions
+
+API property wrappers on parameters have more restrictions than implementation detail property wrappers, because they have an external effect to the function. One very nice thing about property wrappers as custom attributes is that the documentation for that attribute is defined by the library author. Extend this principle, we could allow library authors to also define custom diagnostic notes to help explain restrictions that exist due to the external effects the wrapper imposes on the function.
```

Comment:
```suggestion
API property wrappers on parameters have more restrictions than implementation detail property wrappers, because they have an external effect to the function. One very nice thing about property wrappers as custom attributes is that the documentation for that attribute is defined by the library author. Extending this principle, we could allow library authors to also define custom diagnostic notes to help explain restrictions that exist due to the external effects the wrapper imposes on the function.
```